examples/shared/src/main/scala/zio/cli/examples/WcApp.scala

```diff
@@ -2,27 +2,96 @@ package zio.cli.examples
+    countLines: Option[Long],
+    countWords: Option[Long],
+    countChar: Option[Long]
+  )
 
   val options = (bytesFlag :: linesFlag :: wordsFlag :: charFlag).as(WcOptions)
 
-  val args = Args.file(Exists.Yes).repeat1
+  val args = Args.file("files", Exists.Yes).repeat1
 
   val wc = Command("wc", options, args)
 
-  val execute: (WcOptions, ::[Path]) => URIO[Console, Unit] = (opts, paths) => zio.console.putStrLn(s"${opts} ${paths}")
+  val execute: (WcOptions, ::[Path]) => URIO[Console, Unit] = {
+    def printResult(res: List[WcResult]): ZIO[Console, Nothing, Unit] = {
+      def wcTotal(results: List[WcResult]) = {
+        def optSum(acc: WcResult, elem: WcResult, extract: WcResult => Option[Long]): Option[Long] =
+          extract(acc).flatMap(a => extract(elem).map(_ + a))
+
+        results.fold(WcResult("total", Some(0), Some(0), Some(0), Some(0))) { (acc, wcRes) =>
+          acc.copy(
+            countBytes = optSum(acc, wcRes, _.countBytes),
+            countChar = optSum(acc, wcRes, _.countChar),
+            countWords = optSum(acc, wcRes, _.countWords),
+            countLines = optSum(acc, wcRes, _.countLines)
+          )
+        }
+      }
+      def format(res: WcResult) = {
+        def opt(option: Option[Long]): String = option.map(l => f"${l}%9d").getOrElse("")
+        s"${opt(res.countLines)} ${opt(res.countWords)} ${opt(res.countChar)} ${opt(res.countBytes)} ${res.fileName}"
+      }
+      ZIO.foreach(res)(r => putStrLn(format(r))) *> ZIO.when(res.length > 1)(putStrLn(format(wcTotal(res)))).ignore
+    }
+
+    (opts, paths) => {
+      zio.console.putStrLn(s"executing wc with args: ${opts} ${paths}") *>
+        ZIO
+          .foreachParN[Blocking, Throwable, Path, WcResult, List](4)(paths) { path =>
+            ZStream.fromFile(path).broadcastDynamic(DefaultChunkSize).use { bytesFanOut =>
+              val utf8FanOut = bytesFanOut.map(_.transduce(ZTransducer.utfDecode))
+              val chars      = utf8FanOut.map(_.map(_.length))
+              val words      = utf8FanOut.map(_.transduce(ZTransducer.splitOn(" ")))
+              val lines      = utf8FanOut.map(_.transduce(ZTransducer.splitLines))
+
+              def runCountOrNone(optStream: Option[ZStream[Any, Throwable, _]]) =
+                optStream.map(_.runCount.orDie.optional).getOrElse(ZIO.none)
+              def runSumOrNone[A](optStream: Option[ZStream[Any, Throwable, Int]]): ZIO[Any, Throwable, Option[Long]] =
+                optStream.map(_.runSum.orDie.optional).getOrElse(ZIO.none).map(_.map(_.asInstanceOf[Long]))
+
+              for {
+                bytesStream <- if (opts.bytes) bytesFanOut.optional else ZIO.none
+                charsStream <- if (opts.char) chars.optional else ZIO.none
+                wordsStream <- if (opts.words) words.optional else ZIO.none
+                linesStream <- if (opts.lines) lines.optional else ZIO.none
+
+                bytesFiber <- runCountOrNone(bytesStream).fork
```

Comment:
Unfortunately you will read the file 4 times. I think you want 4 transducers, each of them doing word / character / line / byte count, and then glue them together, and stick them on the stream. That would be much more efficient than reading the file 4 separate times.

Reply:
How do `ZTransducer` compose in that way? There is not `zip` on `ZTransducer`s. But I get your point there. I'll implement it using `ZSink`s. One for each optional count and then zipping them and apply it on the stream.

Reply:
It's unfortunate they don't compose (although they can). So yes, I think switching to using ZSink is a good idea.